pkg/upgrade/v1alpha2/kubernetes.go

```diff
@@ -0,0 +1,442 @@
+}
+
+// Create creates a upgradeTask in specified namespace in kubernetes cluster
+func (k *Kubeclient) Create(
+	upgradeTask *apis.UpgradeTask,
+) (*apis.UpgradeTask, error) {
+	if upgradeTask == nil {
+		return nil, errors.New(
+			"failed to create upgradeTask: nil upgradeTask object",
+		)
+	}
+	cli, err := k.getClientsetOrCached()
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"failed to create upgradeTask {%s} in namespace {%s}",
+			upgradeTask.Name,
+			upgradeTask.Namespace,
+		)
+	}
+	return k.create(cli, k.namespace, upgradeTask)
+}
+
+// Delete deletes a upgradeTask instance from the
+// kubecrnetes cluster
+func (k *Kubeclient) Delete(name string, deleteOpts *metav1.DeleteOptions) error {
+	if strings.TrimSpace(name) == "" {
+		return errors.New("failed to delete upgradeTask: missing upgradeTask name")
+	}
+	cli, err := k.getClientsetOrCached()
+	if err != nil {
+		return errors.Wrapf(
+			err,
+			"failed to delete upgradeTask {%s} in namespace {%s}",
+			name,
+			k.namespace,
+		)
+	}
+	return k.del(cli, k.namespace, name, deleteOpts)
+}
+
+// Patch patches the upgrade task  if present in kubernetes cluster
+func (k *Kubeclient) Patch(
+	name string,
+	pt types.PatchType,
+	data []byte,
+	subresources ...string,
+) (*apis.UpgradeTask, error) {
+	if len(name) == 0 {
+		return nil, errors.New(
+			"failed to patch upgrade task : missing upgradeTask name",
+		)
+	}
+	cli, err := k.getClientsetOrCached()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to patch upgradeTask: {%s}", name)
+	}
+	return k.patch(cli, k.namespace, name, pt, data, subresources...)
+}
+
+// Update updates the upgrade task  if present in kubernetes cluster
+func (k *Kubeclient) Update(
+	upgradeTask *apis.UpgradeTask,
+) (*apis.UpgradeTask, error) {
+	if upgradeTask == nil {
+		return nil, errors.New(
+			"failed to udpate upgradeTask: nil upgradeTask object",
+		)
+	}
+	cli, err := k.getClientsetOrCached()
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"failed to update upgradeTask {%s} in namespace {%s}",
+			upgradeTask.Name,
+			upgradeTask.Namespace,
+		)
+	}
+	return k.update(cli, k.namespace, upgradeTask)
+}
+
+// GetUpgradeDetailedStatuses builds and returns the UpgradeDetailedStatuses
+// object with given parameters
+func GetUpgradeDetailedStatuses(
+	step apis.UpgradeStep,
+	phase apis.StepPhase,
+	// args contains message and reason depending upon the StepPhase
+	args ...string,
+) (*apis.UpgradeDetailedStatuses, error) {
+	uStatusObj := &apis.UpgradeDetailedStatuses{}
+	if step == "" {
+		return nil, errors.Errorf("failed to build upgradestatus : missing step")
+	}
+	uStatusObj.Step = step
+	if phase == "" {
+		return nil, errors.Errorf("failed to build upgradestatus : missing phase")
+	}
+	uStatusObj.Phase = phase
+	uStatusObj.LastUpdatedTime = metav1.Now()
+	if phase == apis.StepWaiting {
+		uStatusObj.StartTime = uStatusObj.LastUpdatedTime
+		return uStatusObj, nil
+	}
+	if len(args) == 0 || args[0] == "" {
```

Comment:
This validation does not go well with variadic args.
If we use variadic then validation should not be there or viceversa

Reply:
Done